libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -0,0 +1,719 @@
+
+        assert_contains!(actual, "      hello world\n");
+    }
+
+    #[test]
+    fn warning_after_buildpack() {
+        let writer = Vec::new();
+        let io = BuildpackOutput::new(writer)
+            .start("RCT")
+            .warning("It's too crowded here\nI'm tired")
+            .section("Guest thoughts")
+            .step("The jumping fountains are great")
+            .step("The music is nice here")
+            .finish()
+            .finish();
+
+        let expected = formatdoc! {"
+
+            # RCT
+
+            ! It's too crowded here
+            ! I'm tired
+
+            - Guest thoughts
+              - The jumping fountains are great
+              - The music is nice here
+            - Done (finished in < 0.1s)
+        "};
+
+        assert_eq!(
+            expected,
+            strip_ansi_escape_sequences(String::from_utf8_lossy(&io))
+        );
+    }
+
+    #[test]
+    fn warning_step_padding() {
+        let writer = Vec::new();
+        let io = BuildpackOutput::new(writer)
+            .start("RCT")
+            .section("Guest thoughts")
+            .step("The scenery here is wonderful")
+            .warning("It's too crowded here\nI'm tired")
+            .step("The jumping fountains are great")
+            .step("The music is nice here")
+            .finish()
+            .finish();
+
+        let expected = formatdoc! {"
+
+            # RCT
+
+            - Guest thoughts
+              - The scenery here is wonderful
+
+            ! It's too crowded here
+            ! I'm tired
+
+              - The jumping fountains are great
+              - The music is nice here
+            - Done (finished in < 0.1s)
+        "};
+
+        assert_eq!(
+            expected,
+            strip_ansi_escape_sequences(String::from_utf8_lossy(&io))
+        );
+    }
+
+    #[test]
+    fn double_warning_step_padding() {
+        let writer = Vec::new();
+        let output = BuildpackOutput::new(writer)
+            .start("RCT")
+            .section("Guest thoughts")
+            .step("The scenery here is wonderful");
+
+        let io = output
+            .warning("It's too crowded here")
+            .warning("I'm tired")
+            .step("The jumping fountains are great")
+            .step("The music is nice here")
+            .finish()
+            .finish();
+
+        let expected = formatdoc! {"
+
+            # RCT
+
+            - Guest thoughts
+              - The scenery here is wonderful
+
+            ! It's too crowded here
+
+            ! I'm tired
+
+              - The jumping fountains are great
+              - The music is nice here
+            - Done (finished in < 0.1s)
+        "};
+
+        assert_eq!(
+            expected,
+            strip_ansi_escape_sequences(String::from_utf8_lossy(&io))
+        );
+    }
+
+    fn strip_ansi_escape_sequences(contents: impl AsRef<str>) -> String {
```

Comment:
Short: I'll add a unit test for warn/important/error/start color codes

Long:

I'm against testing color codes everywhere from personal experience in other libraries. I've said so to Manuel as well. In an ideal state I want the end developer to be able to write a test before any code executes. I also don't want to make the tests so cumbersome that future features or code aren't tested if testing is too cumbersome. I also don't want the state where one change in the style file requires many mechanical changes in other files that will produce PRs with large diffs that obscure the focus.

From this zoom level, the color is a UI concern. In the same way most developers don't assert that a certain style is bold in unit tests, they just make it bold. The color is tested at a lower level. If there's a regression or a problem, then we can revisit what kind of test would best capture that regression and where.

When I test, I test for behavior. I don't really **want** ansi codes, I want colored text, which is covered by earlier tests. I want an easy way to evaluate visual changes as a whole qualitatively. I.e. "Someone changed something in the build output, does it still look okay?" A test won't answer that, but a mechanism that outputs all the various styles for visual inspection will. That was one original purpose of the style guide.

Asserting against color codes yields test output that looks like this:

```
---- buildpack_output::test::test_color_codes stdout ----
thread 'buildpack_output::test::test_color_codes' panicked at libherokubuildpack/src/buildpack_output/mod.rs:558:9:
assertion `left == right` failed
  left: "\n# Heroku Ruby Buildpack\n\n! I am a warning\n\n! I am important\n\n! I am an error\n\n"
 right: "\n\u{1b}[1;35m# Heroku Ruby Buildpack\u{1b}[0m\n\n\u{1b}[0;33m! I am a warning\u{1b}[0m\n\n\u{1b}[1;36m! I am imporant\u{1b}[0m\n\n\u{1b}[0;31m! I am an error\u{1b}[0m\n\n"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

If I REALLY wanted to test this behavior, I would rather want to test that a specific interface was exercised with a specific input. In Ruby I would do this with stubs/mocks.

All of the colored text is within:

- start
- warn
- important
- error

Of these three of them warn/important/error use the same interface. I wouldn't be against a unit test that asserts the behavior of `write_paragraph` given it passes in a specific color. Though that doesn't cover `start`. I'll do that now.

Eh, once I've gone that far, I might as well just assert the actual behavior. I'll add one test that exercises all these 4 interfaces. But as Error consumes, it must use a temp file

Reply:
> I'm against testing color codes everywhere

Yeah I don't think we should test against them everywhere, but we should at least have some tests. (Which we now do, thank you.)

> In an ideal state I want the end developer to be able to write a test before any code executes

I don't understand this phrasing - what does "before any code executes" mean?

> From this zoom level, the color is a UI concern. In the same way most developers don't assert that a certain style is bold in unit tests, they just make it bold. The color is tested at a lower level. If there's a regression or a problem, then we can revisit what kind of test would best capture that regression and where.

I agree, however:
1. For libcnb, the UI **is** a concern, since buildpack maintainers are relying on us to get the output/style/colours right. If we were talking about the tests written in individual buildpacks, then yes, they shouldn't have to test against colours at all.
2. For libcnb, there was a gap in test coverage until the new test was added. (The lower level tests only tested eg ANSI escaping, they didn't actually show the style working as a whole.

Reply:
> I don't understand this phrasing - what does "before any code executes" mean?

TDD style. As in, write the test before you write the code. The harder you make it to write a test, the harder it is to do TDD. I mostly use it as a heuristic that could also approximated as "could someone send me a meaningful PR with a failing test and let me fix it?"

> For libcnb, there was a gap in test coverage until the new test was added. (The lower level tests only tested eg ANSI escaping, they didn't actually show the style working as a whole.)

The original was designed such that testing all the permutations at the highest level wasn't needed, and someone working in this file can choose to trust the style interfaces. 

Previously, the individual style methods were tested, and then those methods were composed into another method that asserted the composition behaved as expected. I didn't realize we had lost some of that coverage when we moved away from the specialized methods (like bangify above).

Reply:
Short: I think we can resolve this.